src/Compression/Compression.cs

```diff
@@ -0,0 +1,415 @@
+		static extern CompressionStatus compression_stream_process (ref CompressionStreamStruct stream, StreamFlag flags);
+
+		[DllImport (Constants.libcompression)]
+		static extern CompressionStatus compression_stream_destroy (ref CompressionStreamStruct stream);
+
+		Stream base_stream;
+		CompressionStreamStruct compression_struct;
+		StreamOperation operation;
+		CompressionAlgorithm algorithm;
+		bool disposed;
+
+		private DeflateStreamNative ()
+		{
+		}
+
+		public static DeflateStreamNative Create (Stream compressedStream, CompressionMode mode, CompressionAlgorithm algorithm)
+		{
+			var dsn = new DeflateStreamNative ();
+			dsn.operation = (mode == CompressionMode.Compress) ? StreamOperation.Encode : StreamOperation.Decode;
+			dsn.algorithm = algorithm;
+			dsn.base_stream = compressedStream;
+			dsn.compression_struct = new CompressionStreamStruct ();
+			
+			var status = compression_stream_init (ref dsn.compression_struct, dsn.operation, dsn.algorithm);
+			if (status != CompressionStatus.Ok)
+				return null;
+			return dsn;
+		}
+
+		~DeflateStreamNative ()
+		{
+			Dispose (false);
+		}
+
+		public void Dispose()
+		{ 
+			Dispose (true);
+			GC.SuppressFinalize (this);
+		}
+
+		public void Dispose (bool disposing)
+		{
+			if (disposed)
+				return;
+
+			compression_stream_destroy (ref compression_struct);
+			disposed = true;
+		}
+
+		public void Flush ()
+		{
+			if (compression_struct.SourceSize != 0) {
+				var closeStatus = compression_stream_process (ref compression_struct, StreamFlag.Finalize);
+				switch (closeStatus) {
+				case CompressionStatus.End:
+					break;
+				case CompressionStatus.Ok:
+					// as per the docs, we should never get here, lets throw an exception so that we know it happened.
+					throw new IOException ("Unexpected CompressionStatus.Ok received.");
+				default:
+					throw new IOException ($"An error occurred when performing the operation: {closeStatus}");
+				}
+			}
+		}
+
+		unsafe public int ReadStream (byte[] array, int offset, int count)
+		{
+
+			var srcManagedBuffer = new byte [count - offset];
+			var dstManagedBuffer = new byte [count - offset];
+			fixed (byte *srcBuf = srcManagedBuffer)
+			fixed (byte *dstBuf = dstManagedBuffer) {
+				var read = base_stream.Read (srcManagedBuffer, offset, count); // we are taking care of the offset in this call
+
+				compression_struct.Destination = (IntPtr) dstBuf;
+				compression_struct.Source = (IntPtr) srcBuf;
+
+				// copy the data to the source buffer from the internal stream
+				compression_struct.DestinationSize = read;
+				compression_struct.SourceSize = read;
+				var readStatus = compression_stream_process (ref compression_struct, StreamFlag.Continue);
+				switch (readStatus) {
+				case CompressionStatus.Ok:
+				case CompressionStatus.End:
+						// copy the data to the passed array, no need to deal with the offset, read took care, count should be read
+						Array.Copy (dstManagedBuffer, 0, array, 0, read);
```

Comment:
I don't quite understand how this works when the output is larger than the input.

For example decompressing data that's 2x compressed:

```csharp
var buffer = new byte [1024];
ReadStream (buffer, 0, buffer.Length);
// now ReadStream will read 1024 bytes from the underlying base stream.
// then ReadStream will call compression_stream_process for those 1024 bytes.
// compression_stream_process will decompress those 1024 bytes into 2048 bytes, but won't be able to write all of those bytes into dstBuf, because it's only 1024 bytes long.
// we copy the first 1024 bytes of decompressed data into dstManagedBuffer and return
// what happens to the second half of the decompressed data? From what I can see, it's lost (the next time ReadStream is called, it will read more data and try to decompress that)
```